services/web/server/src/simcore_service_webserver/catalog_api_handlers.py

```diff
@@ -0,0 +1,376 @@
+        {"data": [json.loads(m.json(by_alias=True)) for m in response_model]}
+    )
+
+    return web.Response(text=enveloped, content_type="application/json")
+
+
+@routes.get(
+    VX + "/catalog/services/{service_key}/{service_version}/outputs/{output_key}"
+)
+@login_required
+@permission_required("services.catalog.*")
+async def get_service_output_handler(request: Request):
+    try:
+        # match, parse and validate
+        service_key: ServiceKey = request.match_info["service_key"]
+        service_version: ServiceVersion = request.match_info["service_version"]
+        output_key: ServiceOutputKey = request.match_info["output_key"]
+
+    except (ValidationError, KeyError) as err:
+        raise web.HTTPUnprocessableEntity() from err
+
+    # Evaluate and return validated model
+    response_model = await get_service_output(
+        service_key, service_version, output_key, _RequestContext.create(request)
+    )
+
+    # format response
+    enveloped: str = json.dumps(
+        {"data": json.loads(response_model.json(by_alias=True))}
+    )
+    return web.Response(text=enveloped, content_type="application/json")
+
+
+@routes.get(VX + "/catalog/services/{service_key}/{service_version}/outputs:match")
+@login_required
+@permission_required("services.catalog.*")
+async def get_compatible_outputs_given_target_input_handler(request: Request):
+    """
+    Filters outputs of this service that match a given service input
+
+    Returns compatible output port of a connected node for a given input
+    """
+    try:
+        # match, parse and validate
+        service_key: ServiceKey = request.match_info["service_key"]
+        service_version: ServiceVersion = request.match_info["service_version"]
+        to_service_key: ServiceKey = request.query["toService"]
+        to_service_version: ServiceVersion = request.query["toVersion"]
+        to_input_key: ServiceInputKey = request.query["toInput"]
+
+    except (ValidationError, KeyError) as err:
+        raise web.HTTPUnprocessableEntity() from err
+
+    # Evaluate and return validated model
+    response_model = await get_compatible_outputs_given_target_input(
+        service_key,
+        service_version,
+        to_service_key,
+        to_service_version,
+        to_input_key,
+        _RequestContext.create(request),
+    )
+
+    # format response
+    enveloped: str = json.dumps({"data": response_model})
+
+    return web.Response(text=enveloped, content_type="application/json")
+
+
+###############
+# IMPLEMENTATION
+#
+
+
+def can_connect(from_output: ServiceOutput, to_input: ServiceInput) -> bool:
+    # FIXME: can_connect is a very very draft version
+
+    # compatible units
+    ok = from_output.unit == to_input.unit
+    if ok:
+        # compatible types TODO: see mimetypes examples in property_type
+        ok = from_output.property_type == to_input.property_type
+        if not ok:
+            ok = "data:*/*" in (from_output.property_type, to_input.property_type)
```

Comment:
Can you elaborate on this logic?

Reply:
This is very basic check as a proof of concept. I think the code is pretty readable: two ports are compatible if
a) have the same type
b) have the same units

In the case of types, the json-schema is still very opened. But, based on the [examples](https://github.com/ITISFoundation/osparc-simcore/blob/master/api/specs/common/schemas/node-meta-v0.0.1.json#L208) the ``data:*/*`` comprises **any** kind of files.